### ISSUE TITLE
Display images only on user request

### ISSUE THREAD

**Issue by mossroy** [kiwix/kiwix-js#173]
Displaying a big article can be quite slow on a low-end device.
It can even crash the browser because of too much memory consumption.
Some optimizations are certainly possible, but not in the short term.

I suspect that the handling of images consumes a lot of resources.
An idea would be to do like most email clients do : first display the content without images, and allow the user to click somewhere to display them. Email clients do that for privacy reasons, we would do that to save resources.
Ideally, an option in the settings would allow the user to indicate if he always wants to display them.

**Comment by Jaifroid** [comment]
@mossroy  Displaying images on user request is now implemented in the Preload stylesheets branch commit https://github.com/kiwix/kiwix-js/commit/634d01d5365135835e87b39bae39fe994504d9e0. See screenshot of configurable options below.

![image](https://user-images.githubusercontent.com/4304337/28513768-4ef5f96e-704f-11e7-8247-c9dad1cfe528.png)

**Comment by Jaifroid** [comment]
@mossroy , @sharun-s : I've now fully the manual download of images in commit https://github.com/kiwix/kiwix-js/commit/6857a40d24012d74e741c605223892a4a2906aea : the code now shows a light yellow area where the image should be, and user can click on it to download just that image. Tested on Firefox, Edge and Edge Mobile. Firefox had an issue whereby the image is rendered un-clickable if no data has been downloaded for it. However, by adding an `alt="Image"` attribute, the image is clickable once again. This means that the clickable area, in Firefox, is only the width of the word "Image". Also, other "alt" text may conflict -- I haven't tested this aspect, as most images don't have alt text, but occasionally some do (if you can think of a better solution, let me know). In the other browsers tested, the clickable area is the actual width of the image to be obtained. Some screenshots from Firefox below.

![image](https://user-images.githubusercontent.com/4304337/28537110-b0b3783c-70a1-11e7-8d69-dbd310330c26.png)

![image](https://user-images.githubusercontent.com/4304337/28537186-ee1d745c-70a1-11e7-92aa-8dc1750e1c96.png)

![image](https://user-images.githubusercontent.com/4304337/28537223-02500a16-70a2-11e7-975f-caf982d8dd3e.png)

![image](https://user-images.githubusercontent.com/4304337/28537252-1255ced2-70a2-11e7-889b-56ff4826f5a5.png)

**Comment by Jaifroid** [comment]
Changed lightyellow to lightblue, as it is less intrusive and fits with the standard link colour.

![image](https://user-images.githubusercontent.com/4304337/28560644-7759d69a-7113-11e7-9a51-fb63c5244cc8.png)

**Comment by mossroy** [comment]
That looks promising!
I quickly tested on Firefox OS and it seems to work well too (after removing the "use strict" because of the inner functions)
A few issues anyway : 
- The checkbox in settings to display images or not does not seem to have any effect : images are always disables
- As you said, the user must click in the center of the image to display it. It's something we need to improve
- Sometimes, the displayed image does not get the correct width/height, it is stretched in a small rectangle without respecting its aspect ratio

**Comment by Jaifroid** [comment]
Helpful report, thanks! I'll sort the inner functions soon. Hmm, the checkbox works for me in UWP, Edge and Firefox, clearly I need to debug it for FFOS. It's the same logic as is used for the jQuery/ServiceWorker radio with the exception that there is a value in "init.js" that sets the default (so this must be where the problem lies). I've noticed some images not recovering their horizontal value properly -- I think it's the image maps. These also don't click through correctly on Wikivoyage. But good to know the code works more or less, as the rest is tweaking, I think.

**Comment by Jaifroid** [comment]
@mossroy Fingers crossed, I think commit https://github.com/kiwix/kiwix-js/commit/19c8a4cf7292e91bf1436618d82bba09441e835f has fixed image display on FFOS (aspect ratio now correct). Also, the tick boxes are now working for me in the FFOS Simulator. I've widened the blue placeholder a little, by using a longer alt text. However, it would be better to have it take the original width of the image, and to do this, I think I need to preload a pixel of the required colour and set it to the width of the image (and a touch-friendly height).

Note that image maps are still looking distorted (elongated, rather than squashed) in all browsers, and this affects Wikivoyage in particular. This is a separate issue to the one corrected by this commit (probably easier to fix, when I get a moment).

When you have a chance, please confirm this commit is working for you on FFOS.

**Comment by mossroy** [comment]
Yes, I confirm the aspect ratio is now OK on Firefox OS, at least for the articles I opened.

**Comment by Jaifroid** [comment]
Light-blue image placeholders now show at the correct width in Firefox and FFOS, with commit https://github.com/kiwix/kiwix-js/commit/378072f5da88707d5098a1ae034d34f99f3703e3 . This is achieved by loading a single pixel of lightBlue into the src attribute of each image, from the local filesystem.

I believe this issue is now resolved and feature-complete, but a PR would have to be part of the overall image rendering system, if it's required for kiwix-js.

**Comment by Jaifroid** [comment]
@mossroy I think there are four options to implement this Issue in the light of the discussion in #504. But each has problems associated with it. The first option is the simplest:

Option 1:

* Don't apply any image manipulations at all to SW mode, as I originally proposed in #504. Users who are likely to need such a feature are using old browsers or underpowered hardware that cannot access SW mode. For example, your FFOS devices will never be able to use SW mode, and nor will IE11, nor Windows Mobile (although upcoming Core OS will replace Windows 10M, sadly it won't be available for current mobile devices). The advantage of this option is that it's simple, and it is the **only** option I can think of that would fully achieve what you state when you say "the only way to make sure we do not mess up the content [in SW mode] is to not touch it at all".

**Comment by Jaifroid** [comment]
Option 2:

* Optionally hide all images in both jQuery and SW mode, and provide a UI **button** to load all of them if the user wants to. There are two disadvantages in the light of what you say in #504:

1. Imagine you have scrolled to a section in the middle of a long article with lots of images. You press the button to see the images you're interested in, and nothing seems to happen because images start loading from the top (or in practice, according to whichever Promise finishes first).
    
2. It doesn't achieve your aim of not touching the DOM in any way in SW mode. First, you have to hide/squash the images in such a way that they don't take up loads of empty vertical space, i.e. you break the layout (IMHO it doesn't matter, because the user can turn it off, but it's an objection you raise). Second, The Service Worker cannot initiate the load of specific DOM elements (no access to the DOM). We have to generate fetch events for each image, and tell SW to stop intercepting them. It is true that the SW can send `Response.redirect()` to reload the whole page, but then all the assets would have to be reloaded (CSS, JavaScript, and they are not cached), and the user would lose their place in the doucment they're reading, plus we break the back button, all of which would IMHO be terrible UX. So, instead we must make the browser think the `src` has changed, so the browser initiates Fetch. In my PR, I did this by adding a small [query string](https://github.com/kiwix/kiwix-js/pull/504/files#diff-c15382f78532323c506b1b89fc90ee29R45) to the `src`. The browser immediately generates a fetch event, and the query string tells SW not to intercept it. ISTM that I will need to do something like this one way or another. Is it acceptable? If not, what alternative can you suggest?

**Comment by Jaifroid** [comment]
Option 3:

* This is similar to Option 2, but instead of loading all images in a random order (determined by the race of Promises), we could load images in the viewport first, wait for them to be loaded, then load the rest "in the background". Viewport prioritization is easy in jQuery mode, so maybe it is best only to apply prioritzation there. In SW mode it is a bit trickier, but if we accept that we have to generate a fetch event for each image anyway (Option 2), it would make sense to generate requests for the viewport first.

**Comment by Jaifroid** [comment]
Option 4:

* Similar to Option 3, but loading of images is automated and only images in the viewport are loaded. Instead of the user having to press a UI button, the loading of hidden images would be based on user action such as scrolling, AKA lazy loading. This option would be off by default, so if the user's browser implements the feature natively, they would never have a reason to turn it on. It is what is implemented in the current PR. From a user perspective, this option is the most natural. No button pressing, no layout changes, works on old browsers and underpowered hardware, and works well in top end browsers too (because images above the fold load first).

**Comment by mossroy** [comment]
Thanks for these detailed explanations.
My preference would be for something like option 2. Let me explain why.
Option 1 would let the user think jQuery mode is better than SW mode because it has more features (and that would be true) : as the SW mode is to be the default one in the future (for all the advantages it has, on supported platforms), it would not ease the transition.
Options 3 and 4 are much better in terms of UX, but they make assumptions on ZIM files content, and DOM manipulations. Loading images in this way is not guaranteed to work on all ZIM files.

Option 2 can be very simple to implement/maintain, even if it's far from being as user-friendly as options 3 and 4. But it's already a big improvement for slow devices.
In jQuery mode, it would load all images (the spinner could be displayed until it's finished, as for the first disadvantage you mentioned)
In SW mode, we must reload the whole page. I know it's not ideal, but it's the only way we can do it in a 100% compatible way with any ZIM file (for example, think of images that could be dynamically added in the `onload` event of the body)

As to how let the user do that, I'm not sure yet.
You mentioned adding a button somewhere : it would be great, but I'm afraid it would waste too much space on the screen for this feature.
In all cases, I would implement it as a general switch (not a temporary one for the current page). Because, in SW mode, we can't know when the page will not request any more URLs (for example, they might be requested upon some further user interaction). So we don't have an easy way to revert the SW setting to "no images" at the right moment, for the next pages.

If putting the general switch always visible on the screen is not possible (or too complicated), we might let it in the settings section.
In this case, it might be relevant to implement another UI improvement we might have discussed : provide the ability for the user to come back to the page he was on, after he temporarily goes to the settings. Even if it involves reloading the whole page from the ZIM file, it's still better than being forced to search for it again. But it's a separate issue (maybe we should open a new one for that)

**Comment by Jaifroid** [comment]
OK, thanks @mossroy . We have #127 for the feature you mention about re-opening the last article after visiting Config. In Kiwix JS Windows this feature is implemented and it's quite simple. We don't need to empty the contents of the iframe, all we need to do is hide the iframe, and show the Config, and you reverse it when you go back. No need to reload the article because we never unloaded it.

You can try that out here: https://kiwix.github.io/kiwix-js-windows/www/ . Load a ZIM, go to a page, then click the Config icon. Then "unclick" the Config icon (i.e. click it again), and the original article should re-appear. NB don't press the Home button to go back, as that will go to the landing page, as in Kiwix JS.

Regarding this issue, I agree that a button just for this feature would be excessive (this is why I made the blue placeholders clickable). Since we can't attach event listeners to the placeholders, I think the only other solution is to require the user to turn the option off explicitly in Config and then reload the current article at least in SW mode.

I'll make a new PR in due course (but currently very busy with exam marking).

**Comment by mossroy** [comment]
It looks great, I think it's the way to go, at least for a first version.

**Comment by Jaifroid** [comment]
Just a quick clarification, @mossroy , when you say "it looks great", are you referring to the custom context menu sample I linked to in #521 that appears referenced just above your comment? Or were you referring to the implementation of hiding and re-showing the article in the iframe when moving between Config / About / (Home) in my last post above? They would be somewhat different implementations.

**Comment by mossroy** [comment]
I was referring to the hiding and re-showing of the article when moving between sections, that you already implemented in kiwix-js-windows

**Comment by MisraAnandita** [comment]
Hi I would like to contribute to this bug. If you could be kind enough to tell me where should I start.
Thank you. I am new to open source and a response would mean a lot to me.

**Comment by mossroy** [comment]
Hi @MisraAnandita , it's cool to see new people willing to contribute.

As you can see with all the comments above, there has been some discussions about how to implement that. And @Jaifroid already proposed an implementation in PR #504, but it was not merged, mainly because I'd prefer something more basic/simple (you can read our discussion in the PR).

Here is what I suggest to implement : 
- add a new setting (next to the existing ones) to display images or not. It should be "on" by default, and this user preference should be kept in a cookie (like for other user preferences). When changing this value, it should warn the user that, on some platforms, the app needs to be restarted for the setting to take effect (it should be the case for Chromium/Chrome when run inside a browser extension, as it uses a built-in memory cache that we can not disable or empty, except by closing/re-opening the extension)
- for the jQuery mode, based on this setting, we should modify function displayArticleContentInIframe (in app.js), so that we call loadImagesJQuery() only if the user wants the images
- for the ServiceWorker mode, based on this setting, we should, in service-worker.js, return an empty content when the requested url looks like an image (i.e, after ignoring the parameters and anchors, if the URL ends with .jpeg, .gif, .png etc. We unfortunately don't know the content-type at this moment, and want to avoid taking the time to find it in the ZIM file). This should be done before looking in the assetCache, around line 81 in the current version of the code

It's clearly not as good as what @Jaifroid previously proposed in terms of UX. I'm aware of that. But it should be much simpler to implement, and should be enough until we have a faster backend.
@Jaifroid : if I missed something, don't hesitate to correct me. If you have other directions, please give them too (like suggestions to avoid code diverging with kiwix-js-windows)

As for how to do that, you'll have to fork this github repo in your own namespace, and create a Pull Request when you have implemented something, so that we can review and discuss it.
To test on your local computer, you have to clone the repo locally. To run the code locally, you have to access index.html through a local webserver (i.e. access it through a URL starting with http://localhost). You can use the webserver you prefer (for example one provided by your favorite code editor, if it does).

If you have questions, we'll try to answer here

**Comment by Hardik500** [comment]
Can I work on this issue if it's still open?

**Comment by Jaifroid** [comment]
@Hardik500 This issue has a complicated history. Please read all the above and look at the code I developed in a PR that was never merged. It is not easy to make this work in a way that is compliant with the Service Worker code. I made a complete solution, but the code added complexity that in the end wasn't felt to be justified given the use case. Maybe you can simplify my code in the way suggested in my comment above this one and in @mossroy's comment above mine.

**Comment by Hardik500** [comment]
@Jaifroid  Ok I will try to simplify your code and will contact you if I get stuck anywhere, if you don't mind.

**Comment by Jaifroid** [comment]
@Hardik500 It might be useful if I rebase the code first on current master (unless you've already started working on it).

**Comment by Hardik500** [comment]
No I haven't started working on it, was just checking the list of tasks that need to be done.

**Comment by Hardik500** [comment]
@Jaifroid sorry for a super late PR, I was busy with my exams. Could you check the PR and see what changes I could improve upon.

**Comment by ykabusalah** [comment]
Is this still open? Could I tackle this?

**Comment by Jaifroid** [comment]
@ykabusalah At the moment @Hardik500 is working on a PR for this issue #605. Is there another issue that interests you?

**Comment by ykabusalah** [comment]
Do you have any other open good first issues that haven't been tackled? I'm currently about to complete one, and I would like another one.

**Comment by ykabusalah** [comment]
@Jaifroid

**Comment by kelson42** [comment]
I wonder about the status and the pertinence of this ticket:
* There is a lot of comments, but the primary problem to fix seems to be performance. But the usage landscape has changed a lot since 5 years. Is performance problem around image loading still a real problem? And a problem we seriously want to fix?
* Lazy image loading via html attribute should be done everywhere meanwhile, so only visible images shoukld be loaded anyway with SW
* Jquery dev should be stalled following our agreement
* Does this ticket has the ambition to treat other problems than the pure performance one? If « yes », then probably new preciser tickets shoukd be open in place of that one
* If a few things have already i pleme ted, as option, we might even consider to remove the option and simplify the configuration landscape?

**Comment by Jaifroid** [comment]
@kelson42 I implemented a version of this that also included lazy loading in jQuery mode (not required as you say in SW mode). However, it was deemed to complicate the codebase too much for little practical gain, precisely because the performance landscape has changed a lot in this app with various types of caching. Some ideas from that PR were incorporated in other PRs. I would say that loading even SVG images is now pretty fast in desktop browsers. There is still an issue on some old mobile platforms, and there is an issue on Android (in the PWA) due to #581. But as I have implemented this PR in Kiwix JS Windows, which is the flavour currently used for the PWA, anyone using the PWA on Android (and there is at least one prolific user who is not me!) can turn off images and load them manually with a click if they want on that implementation. I would say this PR will not be implemented in Kiwix JS, so I close it as you suggest.
